cmd2/cmd2.py

```diff
@@ -2176,120 +2239,352 @@ def _cmdloop(self) -> bool:
-            else:
-                errmsg = "Aliases can not contain: {}".format(invalidchars)
-                self.perror(errmsg, traceback_war=False)
+        utils.unquote_redirection_tokens(args.command_args)
 
-    def complete_alias(self, text: str, line: str, begidx: int, endidx: int) -> List[str]:
-        """ Tab completion for alias """
-        alias_names = set(self.aliases.keys())
-        visible_commands = set(self.get_visible_commands())
+        # Build the macro value string
+        value = args.command
+        for cur_arg in args.command_args:
+            value += ' ' + cur_arg
 
-        index_dict = \
-            {
-                1: alias_names,
-                2: list(alias_names | visible_commands)
-            }
-        return self.index_based_complete(text, line, begidx, endidx, index_dict, self.path_complete)
+        # Find all normal arguments
+        arg_list = []
+        normal_matches = re.finditer(MacroArg.macro_normal_arg_pattern, value)
+        max_arg_num = 0
+        num_set = set()
 
-    @with_argument_list
-    def do_unalias(self, arglist: List[str]) -> None:
-        """Unsets aliases
+        while True:
+            try:
+                cur_match = normal_matches.__next__()
 
-    Usage: unalias [-a] name [name ...]
-    Where:
-        name - name of the alias being unset
+                # Get the number between the braces
+                cur_num = int(re.findall(MacroArg.digit_pattern, cur_match.group())[0])
+                if cur_num < 1:
+                    self.perror("Argument numbers must be greater than 0", traceback_war=False)
+                    return
 
-    Options:
-        -a     remove all alias definitions
-"""
-        if not arglist:
-            self.do_help(['unalias'])
+                num_set.add(cur_num)
+                if cur_num > max_arg_num:
+                    max_arg_num = cur_num
 
-        if '-a' in arglist:
-            self.aliases.clear()
-            self.poutput("All aliases cleared")
+                arg_list.append(MacroArg(start_index=cur_match.start(), number=cur_num, is_escaped=False))
+
+            except StopIteration:
+                break
+
+        # Make sure the argument numbers are continuous
+        if len(num_set) != max_arg_num:
+            self.perror("Not all numbers between 1 and {} are present "
+                        "in the argument placeholders".format(max_arg_num), traceback_war=False)
+            return
+
+        # Find all escaped arguments
+        escaped_matches = re.finditer(MacroArg.macro_escaped_arg_pattern, value)
+
+        while True:
+            try:
+                cur_match = escaped_matches.__next__()
+
+                # Get the number between the braces
+                cur_num = int(re.findall(MacroArg.digit_pattern, cur_match.group())[0])
+
+                arg_list.append(MacroArg(start_index=cur_match.start(), number=cur_num, is_escaped=True))
```

Comment:
It wasn't immediately obvious to me that we're using `start_index=cur_match.start()` to preserve argument order; a comment might be helpful